mmgen/models/ddpm/denoising.py

```diff
@@ -0,0 +1,400 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from copy import deepcopy
+
+import torch
+import torch.nn as nn
+from mmcv.cnn import constant_init
+from mmcv.cnn.bricks.conv_module import ConvModule
+from mmcv.runner import load_checkpoint
+
+from mmgen.models.builder import MODULES, build_module
+from mmgen.utils import get_root_logger
+from .modules import EmbedSequential
+
+
+@MODULES.register_module()
+class DenoisingUnet(nn.Module):
+    """Denoising Unet. This network receives a diffused image ``x_t`` and
+    current timestep ``t``, and returns a ``output_dict`` corresponding to the
+    passed ``output_cfg``.
+
+    ``output_cfg`` define the number of channels and the meaning of the output.
+    ``output_cfg`` support ``mean`` and ``var`` for keys, and
+    denotes how the the network output mean and variance required for denoising
+    process.
+    For ``mean``:
+    1. ``dict(mean='EPS')``: Model would predict noise added in the
+        diffusion process. And the ``output_dict`` would contains a key named
+        ``eps_t_pred``.
+    2. ``dict(mean='START_X')``: Model would direct the mean of the original
+        image `x_0`. And the ``output_dict`` would contains a key named
+        ``x_0_pred``.
+    3. ``dict(mean='X_TM1_PRED')``: Model would predict the mean of diffused
+        image at `t-1` timestep. And the ``output_dict`` would contains a key
+        named ``x_tm1_pred``.
+
+    For ``var``:
+    1. ``dict(var='FIXED_SMALL')`` or ``dict(var='FIXED_LARGE')``: Variance in
+        the denoising process is regarded as a fixed value. The output of
+        network would only have three channels.
```

Comment:
Why do you emphasize the three channels? What's the general case of the output channels?

Reply:
Three channels for  RGB image.